### PR TITLE
feat: partial label expression support

### PR DESCRIPTION
**Pull request by nklaassen**
This commit implements support for label expressions ([RFD 116](https://github.com/gravitational/teleport/blob/master/rfd/0116-label-expressions.md)) with the exception that user traits can not yet be used in the expressions. The user traits will be made available to the expressions in a following PR.

I double-checked that there's no code left that only checks the role labels (without considering the label expression) by temporarily commenting out all the `Get<Kind>Labels` methods in `types.Role` and making sure everything still built. Everything now goes through `role.GetLabelMatchers`. I didn't want to permanently delete the `Get<Kind>Labels` methods because they are part of our public API package.

I added some new tests, and switched a bunch of the roles in `lib/services/role_test.go` to use label expressions instead of plain labels to get extra coverage.

### CODE REVIEW

**Review by fspmarshall — COMMENTED** [review]
I'm concerned about the back-combat story here.  Typically when we add a new feature to roles, it allows something that wasn't previously possible, so it's not a huge deal if an older agent doesn't understand the new field.  Or, worst case, it imposes a new kind of limit (e.g. ip pinning) that it is reasonable to expect might not be consistently enforced until all agents are updated.

This feature could cause older agents to make more permissive decisions than intended.  An older agent might fail to deny an action that would have been denied by the expression, or spuriously treat an action as allowed when the allow expression doesn't actually match.  On its own that isn't so different from new limits not being enforced by old agents, but IMO its more serious than that since we're talking about a top-level access decision.  Furthermore, I wouldn't necessarily expect a user to even *know* that ultimate access is often decided by individual agents rather than the control plane.

There are ways we could work around this problem, though I'm not sure any of them are particularly good.  E.g. we could "censor" roles that use the new feature from the view of older agents somehow... that would cause said agents to fail to perform RBAC checks for users whose roles contain the new feature. super janky, but it *would* work.

We're likely to see issues like this again in the future, so it might be worth tying to think of some more general solution for this kind of problem going forward.  E.g. maybe auth should annotate roles with a "minimum teleport version" based on the features they use, so that going forward agents can know to reject RBAC decisions that they can't reasonably decide on their own.  Idk.  It's a hard problem.

**Comment by nklaassen** [comment]
> I'm concerned about the back-combat story here. Typically when we add a new feature to roles, it allows something that wasn't previously possible, so it's not a huge deal if an older agent doesn't understand the new field. Or, worst case, it imposes a new kind of limit (e.g. ip pinning) that it is reasonable to expect might not be consistently enforced until all agents are updated.
> 
> This feature could cause older agents to make more permissive decisions than intended. An older agent might fail to deny an action that would have been denied by the expression, or spuriously treat an action as allowed when the allow expression doesn't actually match. On its own that isn't so different from new limits not being enforced by old agents, but IMO its more serious than that since we're talking about a top-level access decision. Furthermore, I wouldn't necessarily expect a user to even _know_ that ultimate access is often decided by individual agents rather than the control plane.

Great point, in the past I have just wrote this down as a docs issue and moved on, we're certainly overdue for a better solution here

> There are ways we could work around this problem, though I'm not sure any of them are particularly good. E.g. we could "censor" roles that use the new feature from the view of older agents somehow... that would cause said agents to fail to perform RBAC checks for users whose roles contain the new feature. super janky, but it _would_ work.

The most naive ways to do this may still fail to meet the goal if there's a role with only a deny rule that uses label expressions. Maybe we could return an error if a too-old client attempts to fetch a role that uses the new feature, but of course we want to avoid unexpectedly bricking all nodes as soon as you create a new role. But maybe that would only be a fallback - we could do an inventory check during UpsertRole and block the role creation if any too-old nodes are present. What do you think of that option @fspmarshall? The only other idea I can come up with is inserting wildcard deny labels for the appropriate resource kinds into any roles fetched by too-old nodes, but that is very specific to this one feature

> We're likely to see issues like this again in the future, so it might be worth tying to think of some more general solution for this kind of problem going forward. E.g. maybe auth should annotate roles with a "minimum teleport version" based on the features they use, so that going forward agents can know to reject RBAC decisions that they can't reasonably decide on their own. Idk. It's a hard problem.

Seems like they would just have to reject all RBAC decisions since they'd have no idea what the new feature is?

File: lib/services/role.go [hunk 1]
```diff
@@ -2748,6 +2752,102 @@ func (set RoleSet) checkAccess(r AccessCheckable, state AccessState, matchers ..
...
+		}
+		matches = append(matches, match)
+		messages = append(messages, "expression: "+message)
+	}
+
+	message := strings.Join(messages, " ")
+
+	// Deny rules are greedy, if either matched, it's a match.
+	if condition == types.Deny {
+		return slices.Contains(matches, true), message, nil
```
Comment (greedy52) [inline lib/services/role.go:2828]: nit: wondering how the message looks like. for example:
```
labels: no match, empty selector expression: match
```
It's still readable but probably could improve

File: lib/services/role.go [hunk 2]
```diff
@@ -1853,13 +1852,21 @@ func (set RoleSet) CheckAccessToRemoteCluster(rc types.RemoteCluster) error {
...
+	// For backwards compatibility, if there is no role in the set with label
+	// matchers and the cluster has no labels, assume that the role set has
+	// access to the cluster.
 	usesLabels := false
+outer:
 	for _, role := range set {
-		if len(role.GetClusterLabels(types.Allow)) != 0 || len(role.GetClusterLabels(types.Deny)) != 0 {
-			usesLabels = true
-			break
+		for _, cond := range []types.RoleConditionType{types.Allow, types.Deny} {
```
Comment (greedy52) [inline lib/services/role.go:1861]: nit: this loop is same as `labelMatchersUnset`?

File: lib/services/role_test.go [hunk 3]
```diff
@@ -7630,3 +7937,139 @@ func TestKubeResourcesMatcher(t *testing.T) {
...
+		return role
+	}
+
+	expectDenied := func(tc testcase) bool {
+		return tc.denyLabels == match ||
+			tc.denyLabelsExpression == match ||
+			tc.allowLabels == nomatch ||
+			tc.allowLabelsExpression == nomatch ||
+			(tc.allowLabels == unset && tc.allowLabelsExpression == unset)
+	}
```
Comment (greedy52) [inline lib/services/role_test.go:8030]: just a comment that i think i've spent more than understanding this than reading the actual function =D. normally i wouldn't use logic to validate logic but I understand we are trying to cover all cases possible here.

**Review by greedy52 — APPROVED** [review]
> we could do an inventory check during UpsertRole and block the role creation if any too-old nodes are present. 

m2c on this, this option makes more sense to me. you should not use a feature if your Cluster does not support it.

**Review by greedy52 — APPROVED** [review]
> we could do an inventory check during UpsertRole and block the role creation if any too-old nodes are present. 

m2c on this, this option makes more sense to me. you should not use a feature if your Cluster does not support it.

**Comment by nklaassen** [comment]
> If we make this change part of the v7 changes, that should side-step the problem I think.

It sounds appealing but I'm failing to see what problem that actually solves. I think it just pushes us down a path of trying to downgrade the role for older nodes, which we can already do with or without the role version bump, and I'm not sure if that's the best solution. And since all role versions use the same spec, nothing would implicitly stop a user from using label expressions in a v6 role, unless we explicitly block it, but why would we? There is no breaking change to the role spec here, just new fields, that some older nodes may not see.

The other thing is that it looks like role v7 won't come out until 14.0.0 and label expressions could potentially go out in 13.1.0

Here is my view of our possible options, assuming label expressions would be released in 13.1.0 and role v7 would be released in 14.0.0:

### Option 1 (inventory check and downgrade)
- Write in the docs "don't use label expressions until all nodes run version >v13.1.0"
- UpsertRole will (if label expressions are present) do a best-effort inventory check for services of the relevant Kind on versions older than 13.1.0, and if any are found block the role creation with a helpful error message
- GetRole and equivalents will downgrade the role for any clients on versions < 13.1.0. To "downgrade" a role is to insert wildcard deny labels for all resource kinds with any label expression set. A message will be logged at DEBUG level.
- Users may be confused why they can't access any <v13.1.0 nodes that weren't connected when they upserted the role with the label expressions

### Option 2 (just downgrade)
- Write in the docs "don't use label expressions until all nodes run version >v13.1.0"
- GetRole and equivalents will downgrade the role for any clients on versions < 13.1.0. To "downgrade" a role is to insert wildcard deny labels for all resource kinds with any label expression set. A message will be logged at DEBUG level.
- Users may be confused why they can't access any <v13.1.0 nodes after they create a role with label expressions

### Option 3 (wait for role v7)
- Document that label expressions only work in v7 roles, and won't work for any nodes <14.0.0
- UpsertRole will block label expressions in role versions < v7
- GetRole and equivalents will downgrade the role for any clients on versions < 14.0.0. To "downgrade" a role is to insert wildcard deny labels for all resource kinds with any label expression set. A message will be logged at DEBUG level.
- Users may be confused why they can't access any <v14.0.0 nodes after they create a role with label expressions
- Users may be confused why they have to upgrade their role to v7 and what the implications of that are

### Option 4 (just document it)
- Write in the docs "don't use label expressions until all nodes run version >v13.1.0"
- If this warning is ignored, we may erroneously allow or deny access to any <v13.1.0 nodes that should be denied based on label expressions
- Users may be confused why access decisions for <v13.1.0 nodes are incorrect

I think option 1 is best, and options 2 and 3 are almost equivalent in terms of UX, but option 3 makes us wait for v14. @fspmarshall do you still think role v7 is best or do you have any other thoughts on this?

File: lib/services/role.go (hunk 2 above)
Comment (nklaassen) [reply lib/services/role.go]: right, good catch, thanks

File: lib/services/role_test.go (hunk 3 above)
Comment (nklaassen) [reply lib/services/role_test.go:8030]: Yeah in general I agree, I started off by writing individual test cases with hard-coded expected results, but quickly realized I was missing many edge cases. After simplifying the test cases so I could easily repeat them to get the coverage I wanted, I realized they were very repetitive and I was still missing edge cases, so decided to generate the test cases to get full coverage. Writing this did actually help me to distill the logic and catch some errors in my first implementation.

This snippet is very information-dense so it makes sense that it requires a careful read, but it actually defines the entire behaviour of `checkRoleLabelsMatch` and I think it's kind of nice to have it expressed in a single boolean expression.

File: lib/services/role.go (hunk 1 above)
Comment (nklaassen) [reply lib/services/role.go:2829]: note: we'll never get the "empty selector" message anymore because we check for empty in this function, it will either be `no key match: 'env'` or `no value match: got 'prod' want: '[staging]'`

Made some minor improvements, the message will now look like `windows_desktop_service \"\" denied, deny rule in role \"rolename\" matched; match(namespace=matched, label=no value match: got 'prod' want: '[staging]', expression=matched)"`

**Comment by fspmarshall** [comment]
Sorry for the delayed reply!

> option 3 makes us wait for v14

Ah, I hadn't realized we weren't backporting the new role version to v13.  That does complicate things.

In that case, I suppose option 1 is probably the best we can do with what we've go.

> We've typically addressed this with a warning in the documentation - see the [locking documentation](https://goteleport.com/docs/access-controls/guides/locking/#version-warning)

I don't think this is a perfect comparison. Risk-mitigation mechanisms like locking/ip pinning/etc are applied to what are fundamentally permissible actions.  Without special handling, label expressions would introduce a scenario in which an action might be permitted by an older agent that should never _under any circumstances_ be allowed.  That's a pretty big difference IMO.

**Comment by greedy52** [comment]
another thought on inventory check, nothing really prevents a new older agent to join after UpsertRole

**Comment by nklaassen** [comment]
> another thought on inventory check, nothing really prevents a new older agent to join after UpsertRole

yeah, an inventory check could only ever be a best-effort attempt to give the user an early warning. We can still downgrade the role exactly [the same way we will for v7 roles](https://github.com/gravitational/teleport/blob/c00a5fcc6ae41f9bdbda14569333839df7bec878/lib/auth/grpcserver.go#L1934) to make sure older clients will never errantly allow access. But we don't need to wait for v7 roles to do that, we can downgrade based on the client version alone while still shipping this in v6 roles

**Comment by nklaassen** [comment]
> I suppose option 1 is probably the best we can do

@fspmarshall I've implemented this in a followup PR https://github.com/gravitational/teleport/pull/27244

> We could backport Role v7/6.1 to v13 if it simplifies things. The worst part is that we need to set the downgrade version to compare against a patch releasev13.0.3 instead of a major release which we usually do.

@tigrato I don't think putting this in v7/6.1 actually helps because this isn't changing any defaults, there is no field that would have a different meaning between v6 and v7 roles. We can still do the downgrade logic for certain features within v6 without bumping the role version, as I've implemented in https://github.com/gravitational/teleport/pull/27244

**Review by fspmarshall — APPROVED** [review]
LGTM once https://github.com/gravitational/teleport/pull/27244 gets in.

I think this is definitely the kind of change where some extra manual testing post-merge would be good.  Especially manually testing the back-compat scenarios.
